### ISSUE TITLE
Ensure CurrentIdentityAssociation.setIdentity() refreshes injected TokenCredential's

### ISSUE THREAD

**Issue by wabrit** [quarkusio/quarkus#17746]
## Description

The `CurrentIdentityAssociation.setIdentity(SecurityIdentity)` method ensures that request-scoped injected `SecurityIdentity`'s are refreshed to pick up the new value, but has no similar effect on injected `TokenCredential`'s.

Implementing this would help to ensure more consistent effects to request-scoped security elements in the context, and to allow callers the ability to drive the use of different tokens in a number of scenarios e.g.

- Temporarily running code under elevated privilege ("runAsSystem") using a token obtained via `OidcClient`.
- Directing `@AccessToken`-annotated REST clients to use a different token without having to expose different REST clients to the same remote service for different security scenarios, or adding token-valued parameters to methods of those clients which can be bound to the HTTP `Authorization` request header.
- Writing service methods that are agnostic as to whether they are called from incoming REST requests (where tokens are natively propagated from that request) or from e,g, Kafka/AMQP event handlers (where a token may need to be manually requested via `OidcClient`); such methods could e.g. create audit records based on token information without needing to be aware of how the current token in the context was obtained.

## Implementation ideas

None proposed.

**Comment by sberyozkin** [comment]
See also #12650.
CC @stuartwdouglas @mkouba

**Comment by sberyozkin** [comment]
@michalvavrik Michal, please think about it if you think it is doable

**Comment by michalvavrik** [comment]
> @michalvavrik Michal, please think about it if you think it is doable

Yes, this is doable, but I think it is implementation detail that `SecurityIdentity` proxy is updatable (though honestly I have no means to know this). Ideally, when CDI request context is activated, the identity should be unmodifiable. Just imagine where would what is proposed in the description lead?

- you access `SecurityIdentity` from Arc container and get from it ID (or principal, whatever) that you hold in variable `String principal = identity.getPrincipal().getName()`
- some code is happening downstream, e.g. `callSvcB()` and `CurrrentIdentityAssociation` is changed to a new `SecurityIdentity`
-  when `callSvcB()` finished and you would want to know `principal`, you would be obliged to use `SecurityIdentity` proxy again, otherwise you can't ensure consistency
- but you can't ever enforce noone storing information about existing identity just few code lines above

Honestly I'd go in opposite direction, but I don't believe it will be accepted as it would require breaking changes. If users need to run as different user for a while, I'd do it on a duplicated context with new CDI request context. Problem is how to filter context so that only beans you wish sticks there and others are re-created.

@sberyozkin please if you see something here I don't, let me know. I think this is dangerous for users to adjust `io.quarkus.security.identity.CurrentIdentityAssociation#setIdentity(io.quarkus.security.identity.SecurityIdentity)`. Now, extensions are doing that, but there is a good chance they are doing it before any application code is executed.

I vote for closing the issue.
